README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# minihealth

<!-- badges: start -->
<!-- badges: end -->

The `minihealth` package is a central component of
[JAMES](https://github.com/stefvanbuuren/james). The package

-   Defines S4 classes `xyz`, `bse`, `ird`, `individual` and `cabinet`
    for analysing individual growth data;
-   Translates `individual`, `donordata` and `bds` formats into each
    other;
-   Extracts data and ranges from `individual` objects.

## Installation

Install the development version `minihealth` by

``` r
install.packages("remotes")
remotes::install_github("stefvanbuuren/minihealth")
```

There is no release on CRAN.

## Example 1: Automatic Z-score calculation

The S4 class `xyz` stores three variables useful for anthropometric
data:

-   `x`: usually age, but can also be height (in weight-for-height);
-   `y`: measurement, e.g. height or weight;
-   `z`: Z-score of `y` conditional on `x`.

Here are some examples for automatic *Z*-score calculation:

``` r
library(minihealth)
#> Loading required package: nlreferences
#> Loading required package: donorloader

# specify length (in cm) for boy at ages 0, 0.2 and 0.5 years
new("xyz", x = c(0, 0.2, 0.5), y = c(51.0, 54.1, 63.4))
#>   age  hgt hgt_z
#> 1 0.0 51.0    NA
#> 2 0.2 54.1    NA
#> 3 0.5 63.4    NA

# at the minimum, specify sex for automatic Z-score calculation
new("xyz", x = c(0, 0.2, 0.5), y = c(51.0, 54.1, 63.4), sex = "male")
#>   age  hgt  hgt_z
#> 1 0.0 51.0 -0.154
#> 2 0.2 54.1 -2.312
#> 3 0.5 63.4 -1.829

# specify weight (in kg) at same ages
d1 <- new("xyz", x = c(0, 0.2, 0.5), y = c(3.2, 5.2, 7.0), 
          sex = "male", yname = "wgt")
d1
#>   age wgt  wgt_z
#> 1 0.0 3.2 -0.874
#> 2 0.2 5.2 -0.575
#> 3 0.5 7.0 -1.048

# View reference names used to calculate Z-scores in d1
data.frame(d1)
#>   xname yname zname   x   y      z          pkg             refcode
#> 1   age   wgt wgt_z 0.0 3.2 -0.874 nlreferences nl_1997_wgt_male_nl
#> 2   age   wgt wgt_z 0.2 5.2 -0.575 nlreferences nl_1997_wgt_male_nl
#> 3   age   wgt wgt_z 0.5 7.0 -1.048 nlreferences nl_1997_wgt_male_nl
```

It is also possible to perform the reverse calculation, where `z` is
given and `y` is calculated.

``` r
# Standard weight centiles at age 0.5 year of Dutch boys
new("xyz", x = rep(0.5, 5), z = -2:2, sex = "male", yname = "wgt")
#>   age  wgt wgt_z
#> 1 0.5 6.24    -2
#> 2 0.5 7.04    -1
#> 3 0.5 7.90     0
#> 4 0.5 8.83     1
#> 5 0.5 9.82     2

# Extend to grid of ages: 0y, 0.5y and 1y
new("xyz", x = rep(c(0, 0.5, 1), each = 5), z = rep(-2:2, 3), 
    sex = "male", yname = "wgt")
#>    age   wgt wgt_z
#> 1  0.0  2.77    -2
#> 2  0.0  3.15    -1
#> 3  0.0  3.55     0
#> 4  0.0  3.97     1
#> 5  0.0  4.40     2
#> 6  0.5  6.24    -2
#> 7  0.5  7.04    -1
#> 8  0.5  7.90     0
#> 9  0.5  8.83     1
#> 10 0.5  9.82     2
#> 11 1.0  8.14    -2
#> 12 1.0  9.14    -1
#> 13 1.0 10.24     0
#> 14 1.0 11.44     1
#> 15 1.0 12.75     2
```

See the help (“?`xyz-class`”) and the
[centile](https://github.com/growthcharts/centile) package for more
examples.

## Example 2: Automatic brokenstick estimation

``` r
# specify three height measures
boy <- new("xyz", x = c(0, 0.2, 0.5), y = c(51.0, 54.1, 63.4), 
           sex = "male")
boy
#>   age  hgt  hgt_z
#> 1 0.0 51.0 -0.154
#> 2 0.2 54.1 -2.312
#> 3 0.5 63.4 -1.829

# calculate broken stick estimates at observed ages
new("bse", data = boy)
#> donorloader::load_data(dnr = "smocc_bs", element = "hgt")
#>   age  hgt  hgt_z
#> 1 0.0 50.6 -0.349
#> 2 0.2 54.9 -1.982
#> 3 0.5 63.5 -1.801

# calculate broken stick estimates at all break points
new("bse", data = boy, at = "knots")
#> donorloader::load_data(dnr = "smocc_bs", element = "hgt")
#>       age  hgt  hgt_z
#> 1  0.0000 50.6 -0.349
#> 2  0.0767 50.9 -1.640
#> 3  0.1533 53.2 -1.945
#> 4  0.2500 56.5 -2.020
#> 5  0.3333 60.0 -1.599
#> 6  0.5000 63.5 -1.801
#> 7  0.6250 66.5 -1.532
#> 8  0.7500 68.1 -1.740
#> 9  0.9167 70.7 -1.708
#> 10 1.1667 74.3 -1.605
#> 11 1.5000 78.3 -1.588
#> 12 2.0000 84.0 -1.483
#> 13 3.0000 92.4 -1.547
```

See the help (“?`bse-class`”) and the
[brokenstick](https://github.com/growthcharts/brokenstick) package for
more examples.

## Example 3: Bundle all measures by individual

The S4 class `individual` bundles four different types of information,
each of which is coded by its own class:

-   `individualID`: ID information, like name, date of birth;
-   `individualBG`: Background, like sex, gestional age or etnicity;
-   `individualAN`: Bundles person’s anthrometric data, like `hgt` and
    `wgt`;
-   `individualBS`: Bundles person’s brokenstick estimates;
-   `individualRW`: Stores raw data, e.g. milestones;

Creating an instance of class `individual` can be done by hand in two
steps. First, create one or more of the four subclasses, and then  
tie these together, as follows:

``` r
pid <- new("individualID", name = c("Rob", "Dorchester"),
           dob = as.Date("2014-08-22", "%Y-%m-%d"), 
           id = as.integer(204))
pbg <- new("individualBG", sex = "male", hgtf = 185)
pan <- new("individualAN",
           hgt = new("xyz", 
                     x = c(0, 0.2, 0.5), 
                     y = c(51.0, 54.1, 63.4), 
                     sex = pbg@sex),
           wgt = new("xyz", 
                     x = c(0, 0.5), 
                     y = c(3.2, 7.0), 
                     yname = "wgt", 
                     sex = pbg@sex))
pbs <- new("individualBS",
           bs.hgt = new("bse", yname = "hgt",
                        data = pan@hgt,
                        at = "knots",
                        sex = pbg@sex),
           bs.wgt = new("bse", yname = "wgt", 
                        data = pan@wgt, 
                        at = "knots",
                        sex = pbg@sex))
data <- data.frame(age = c(0.2, 0.5, 0.7),
                   k1430 = c(1, NA, NA),
                   k1431 = c(2, NA, NA),
                   k1437 = c(3, 1, 1),
                   k1438 = c(0, 1, 1),
                   k1439 = c(0, 1, 1))
map <- data.frame(from = c("k1430", "k1431", "k1437", "k1438", "k1439"),
                  to = c(879, 927, 928, 881, 883))
prw <- new("individualRW",
           ddi = new("ird", mst = data, map = map, instrument = "ddi"))

rob <- new("individual", pid, pbg, pan, pbs, prw)
```

Doing this sequence by hand is somewhat inconvenient. Fortunately, there
are two functions that convert other formats into an object of class
`individual`:

-   `donordata_to_individual()` takes data in `donordata` format, and
    converts it into `individual` format;
-   `convert_bds_individual()` takes data in `bds` format, and converts
    it into `individual` format;

See the respective documentation for more detail. Both functions have
inverse (but lossy) transformations.

## Miscellaneous functionality

-   A `cabinet` is a collection of multiple objects of class
    `individual`;
-   `data.frame(d1)` extracts the data frame from objects of class
    `xyz`;
-   `get_xyz(rob, "hgt")` extracts the data frame from objects of class
    `individual`;
-   `get_range(rob)` extract the age range from objects of class
    `individual`.


NEWS.md
# minihealtg 0.91.0

* Prevents `pid@src` from becoming empty
* Adds tests for `individual_to_donordata()`
* Adds separate birth weight field to weight table

# minihealth 0.90.0

* Major update: Replaces `clopus` package by `centile` and `nlreferences`

# minihealth 0.86.1

* Removes all `lollypop.preterm` pointers

# minihealth 0.86.0

* Explicit declaration of `clopus::y2z()` and `clopus::z2y()`

# minihealth 0.85.0

* Repairs calculation of DAZ
* Repair range check, and setting to NA, for out-of-range GA
* Re-edit to `lollypop` to account for `donordata 0.20.0`

# minihealth 0.84.0

* Renames ".z" to "_z"

# minihealth 0.83.0 

* Extends `xyz` class to weight-for-height Z-scores

# minihealth 0.82.0 

* Changes the transformations in `donordata_to_indivual()` and 
`convert_bds_individual()` to use clopus::transform_y()` and 
`clopus::transform_z()` functions.

# minihealth 0.81.0 

* The `xyz` and `bse` classes now support `clopus::transform_z()` and  
`clopus::transform_y()` transformations. Add `usetransform = TRUE, sex = .., ga = ..`
arguments to the `new("xyz", ...)` and `new("bse", ...)` calls to activate this
option.

# minihealth 0.80.2 

* Support only built-in schema

# minihealth 0.80.1

* Add fast return in `convert_bds_individual()` for `NULL` input

# minihealth 0.80.0

* Adds the `convert_checked_individual()` workhorse

# minihealth 0.79.0

* Adds `verify()` to isolate JSON testing from conversion to `individual`

# minihealth 0.78.0

* Replace `individualMS` by more general `individualRW` for raw data storage.

# minihealth 0.77.0

* Repairs six incorrect BDS numbers in `inst/json/bds_schema_str.json`.
* Updates `data/bds_lexicon` with the six repairs.

# minihealth 0.76.0

* Solves the pesky warning message: "Unreplaced values treated as NA as .x is not compatible. Please specify replacements exhaustively or supply .default"
* Turns most warnings into messages. The only warnings remaining occur if the input JSON is invalid.

# minihealth 0.75.0

* Update to `R 4.0.0` and `brokenstick 0.70.1`

# minihealth 0.74.1

* Add bug fix that creates a synthetic dob (20000101) when NA

# minihealth 0.74.0

* preserves the GA in days by extending the `individualBG` class with a `gad` slot
* adds BDS 989 (throws ball) to the JSON validator schema

# minihealth 0.73.0

* `convert_individual.bds()` includes the raw ddi data to BDS JSON

# minihealth 0.72.0

* Replaces calls `as(..., "data.frame")` by `data.frame(...)`

# minihealth 0.71.0

* Generalises the `as(..., "data.frame")` functions for `xyz` and `individualAN` so that the produced data frames are complete and tidy

# minihealth 0.70.0

* Adds functionality for milestone scores. Two new classes: `individualMS` and `ird`
* Extends the `individual` class
* Updates to new classes in `convert_bds_individual()` and `donordata_to_individual()`

# minihealth 0.69.0

* Adds class `individualMS` to convert raw milestones data (from SMOCC) into BDS coding

# minihealth 0.68.1

* Avoid `Unknown or uninitialised column:` warning generated by `tibble`
* Round Z-scores to 3 digits after decimal point

# minihealth 0.46.0

* Export `get_dob()`

# minihealth 0.38.2

* Reduces dependency on `donordata` by importing functions from `donorloader`
* Added a `NEWS.md` file to track changes to the package.


R/class-xyz.R
#' An S4 class for anthropometric data
#'
#'The \code{xyz} class stores three variables useful for growth curves:
#'\code{x} (usually age), \code{y} (height, weight, etc) and \code{z} (Z-score,
#'or standard deviation score (SDS)
#'corresponding to \code{y} given \code{x}). The class calculates
#'\code{z} (when \code{y} is specified), or \code{y} (when \code{z} is specified).
#'
#'@section Slots:
#'  \describe{
#'    \item{\code{x}:}{Numeric vector of decimal ages (in years) at which
#'    measurements \code{y} and Z-scores \code{z} are made. The length
#'    of \code{x} determines how many measurements (including \code{NA}'s)
#'    are stored.}
#'    \item{\code{y}:}{Numeric vector of measurement of length \code{length(x)}.}
#'    \item{\code{z}:}{Numeric vector of Z-score of length \code{length(x)}.}
#'    \item{\code{xname}:}{A character scalar with the name of the \code{x}
#'    variable,usually something like \code{"age"}.}
#'    \item{\code{yname}:}{A character scalar with the name of the measurement.
#'    Currently restricted to one of \code{c("hgt", "wgt", "hdc", "wfh", "bmi", "dsc")}
#'    for height (in cm), weight (in kg), head circumference (in cm),
#'    weight-for-height (kg/m), body mass index (kg/m^2) and D-score, respectively.}
#'    \item{\code{zname}:}{A character scalar with the name of the \code{z}
#'    variable. By default, is it equal to \code{paste0(yname, "_z")}.}
#'    \item{\code{call}:}{An object of class \code{call} that specifies the
#'    appropriate reference table from the \pkg{clopus} package. Use the function
#'    \code{\link[clopus]{create.reference.call}} to set \code{call} manually.
#'    Alternatively, specify one or more of \code{libname}, \code{prefix},
#'    \code{sex},
#'    \code{yname} and \code{sub} arguments, which are passed down to
#'    \code{create.reference.call()}. DEPRECATED.}
#'    \item{\code{refcode}:}{Since \code{minihealth 0.90.0}, Name of the reference. See
#'    \code{\link[centile]{make_refcode}}.}
#'    \item{\code{pkg}:}{Since \code{minihealth 0.90.0}, Name of the package containing
#'    the reference. See \code{\link[centile]{make_refcode}}. The default is
#'    \code{nlreferences}.}
#'    \item{\code{found}:}{Internal logical flag indicating
#'    whether the reference specified by
#'    \code{call} was actually found.}
#'    \item{\code{usetransform}:}{Logical indicating whether conversion to
#'    and from should be done using \code{nlreferences::transform2z()} and
#'    \code{nlreferences::transform2y()}.}
#'    \item{\code{transform}:}{Name of the \code{nlreferences} transform function.}
#'    \item{\code{sex}:}{Sex specified by user argument.}
#'    \item{\code{ga}:}{Gestational age specified by user argument.}
#'    \item{\code{age}:}{Decimal age specified by user argument.}
#'    \item{\code{digits}:}{Number of digits for \code{round()} function. Default
#'    is 3.}
#'}
#'
#'@name xyz-class
#'@rdname xyz-class
#'@aliases xyz-class
#'@author Stef van Buuren
#'@keywords classes
#'@examples
#'# specify length (in cm) for boy at ages 0, 0.2 and 0.5 years
#'d1 <- new("xyz", x = c(0, 0.2, 0.5), y = c(51.0, 54.1, 63.4))
#'d1
#'
#'# Z-scores of weight (in kg) at same ages
#'d2 <- new("xyz", x = c(0, 0.2, 0.5), y = c(3.2, 5.2, 7.0), yname = "wgt",
#'   sex = "female")
#'d2
#'
#'# As as data.frame
#'data.frame(d2)
#'
#'# Head circumference of girl, relative to WHO standard
#'d3 <- new("xyz", x = c(0, 0.2, 0.5), y = c(35, 38, 41), pkg = "centile",
#' refcode = "who_2007_hdc_female_")
#'d3
#'data.frame(d3)
#'
#'# Standard weight centiles at age 0.5 year of Dutch girls
#'d5 <- new("xyz", x = rep(0.5, 5), z = -2:2, sex = "female", yname = "wgt")
#'d5
#'
#'# calculate centiles at 1 year using the female WHO head circumference reference
#'d6 <- new("xyz", yname = "hdc", x = rep(1, 5), z = -2:2, pkg = "centile",
#' refcode = "who_2007_hdc_female_")
#'d6
#'
#'# calculate P50 for preterms, female and born a GA week 32
#'d7 <- new("xyz", yname = "hgt", x = seq(0, 1, 1/12), z = rep(0, 13),
#'  refcode = "nl_2012_hgt_female_32")
#'d7
#'
#'# use nlreferences::transform2y() as an alternative
#'d8 <- new("xyz", yname = "hgt", x = seq(0, 1, 1/12), z = rep(0, 13),
#'   usetransform = TRUE, ga = 32, sex = "female")
#'d8
#'
#'# use nlreferences::transform2z function
#'d9 <- new("xyz", yname = "hgt", x = seq(0, 0.25, 1/12), y = c(43, 46, 48, 50),
#'   usetransform = TRUE, ga = 32, sex = "female", digits = 10)
#'d9
#'@export
setClass("xyz",
         slots = c(
           x = "numeric",
           y = "numeric",
           z = "numeric",
           xname = "character",
           yname = "character",
           zname = "character",
           refcode = "character",
           pkg = "character",
           usetransform = "logical",
           transform = "character",
           sex = "character",
           ga = "numeric",
           age = "numeric",
           digits = "integer",
           call = "language",  # deprecated
           found = "logical"   # deprecated
         ), prototype = list()
)

setMethod(
  "initialize", "xyz",
  function (.Object, x = numeric(0), y, z,
            xname = "age", yname = "hgt",
            refcode = character(0),
            pkg = "nlreferences",
            usetransform = FALSE,
            verbose = FALSE,
            sex = NA_character_,
            ga = NA_real_,
            age = NA_real_,
            digits = 3L,
            call = NULL,
            ...) {

    if (!missing(call)) {
      stop("Argument `call` is deprecated; please use `refcode` and `pkg` instead.",
           call. = FALSE)
    }

    # store flag
    slot(.Object, "usetransform") <- usetransform
    slot(.Object, "transform") <- "not set"

    # filter sex, ga and age from dots
    # catch <- function(sex = NA_character_,
    #                   ga = NA_real_,
    #                   age = NA_real_, ...) {
    #   list(sex = sex, ga = ga, age = age)
    # }
    # sexga <- catch(...)
    .Object@sex <- sex
    .Object@ga  <- ga
    .Object@age <- age
    .Object@digits <- as.integer(digits)

    # direct refcode/pkg specification overrides everything else
    # else, create automatic refcode from the data
    if (!missing(refcode)) {
      slot(.Object, "refcode") <- refcode
    } else {
      df <- data.frame()
      if (length(x)) {
        df <- data.frame(xname = xname,
                         yname = yname,
                         x = as.numeric(x),
                         sex = .Object@sex,
                         age = .Object@age,
                         ga = .Object@ga)
      }
      slot(.Object, "refcode") <- nlreferences::set_refcodes(df)
    }
    slot(.Object, "pkg") <- pkg

    # obtain reference table
    #ref <- centile::load_reference(refcode,
    #                               pkg = pkg,
    #                               verbose = verbose)
    #.Object@found <- !is.null(ref)

    # initialize x
    slot(.Object, "x") <- as.numeric(x)
    lx <- length(slot(.Object, "x"))

    # initialize y
    if (missing(y)) {
      if (missing(z)) slot(.Object, "y") <- rep(NA_real_, lx)
      else if (usetransform) {
        slot(.Object, "transform") <- "transform2y()"
        df <- data.frame(z = as.numeric(z),
                         x = slot(.Object, "x"),
                         sex = .Object@sex,
                         ga = .Object@ga)
        names(df) <- c(paste0(yname, "_z"), "age", "sex", "ga")
        slot(.Object, "y") <-
          nlreferences::transform2y(df, znames = paste0(yname, "_z"),
                                    verbose = verbose)[[yname]]
      }
      else {
        .Object@y <- centile::z2y(z = as.numeric(z),
                                  x = .Object@x,
                                  refcode = .Object@refcode,
                                  pkg = .Object@pkg,
                                  verbose = verbose,
                                  dec = .Object@digits,
                                  ...)
      }
    }
    else slot(.Object, "y") <- as.numeric(y)
    if (length(slot(.Object, "y")) == 0)
      slot(.Object, "y") <- rep(NA_real_, lx)

    # initialize z
    if (missing(z)) {
      if (missing(y)) slot(.Object, "z") <- rep(NA_real_, lx)
      else if (usetransform && lx) {
        slot(.Object, "transform") <- "transform2z()"
        if (yname != "wfh") {
          df <- data.frame(y = slot(.Object, "y"),
                           x = slot(.Object, "x"),
                           sex = .Object@sex,
                           ga = .Object@ga)
          names(df) <- c(yname, "age", "sex", "ga")
        } else {
          df <- data.frame(y = slot(.Object, "y"),
                           x = slot(.Object, "x"),
                           sex = .Object@sex,
                           ga = .Object@ga,
                           age = .Object@age)
          names(df) <- c("wgt", "hgt", "sex", "ga", "age")
        }
        slot(.Object, "z") <-
          nlreferences::transform2z(df, ynames = yname,
                                    verbose = verbose)[[paste0(yname, "_z")]]
      }
      else {
        .Object@z <- centile::y2z(y = as.numeric(y),
                                  x = .Object@x,
                                  refcode = .Object@refcode,
                                  pkg = .Object@pkg,
                                  verbose = verbose,
                                  dec = .Object@digits,
                                  ...)
      }
    }
    else slot(.Object, "z") <- round(as.numeric(z), .Object@digits)
    if (!length(slot(.Object, "z")))
      slot(.Object, "z") <- rep(NA_real_, lx)

    # administrative names
    .Object@xname <- as.character(xname[1L])
    .Object@yname <- as.character(yname[1L])
    if (!length(.Object@zname)) .Object@zname <- paste0(yname[1L], "_z")

    check <- validObject(.Object)
    return(.Object)
  }
)

setValidity("xyz", function(object) {
  if (length(object@y) > 1L && length(object@x) != length(object@y))
    return(paste0("Non-conformable length: ",
                  object@xname, " (", length(object@x),") and ",
                  object@yname, " (", length(object@y),")."))
  if (length(object@z) > 1L && length(object@x) != length(object@z))
    return(paste0("Non-conformable length: ",
                  object@xname, " (", length(object@x),") and ",
                  object@zname, " (", length(object@z),")."))
  if (length(object@xname) > 1L) return("xname not of length 1")
  if (length(object@yname) > 1L) return("yname not of length 1")
  if (length(object@zname) > 1L) return("zname not of length 1")
  return(TRUE)
})

setMethod("show", signature(object = "xyz" ),
          function (object) {
            if (object@usetransform) {
              cat(paste0("Transform ", object@pkg, "::", object@transform, "\n"))
            }
            if (length(object@x) | length(object@y) | length(object@z)) {
              df <- data.frame(object@x, object@y, object@z)
              names(df) <- c(object@xname, object@yname, object@zname)
              print(df)
              return()
            }
          cat("No data for ", object@xname, ", ", object@yname,
              " and ", object@zname, "\n", sep = "")
          }
)

#' as("xyz", "data.frame")
#'
#' @name as
#' @family xyz
setAs("xyz", "data.frame", function(from) {
  if (length(from@x) && from@usetransform)
    return(data.frame(xname = from@xname,
                      yname = from@yname,
                      zname = from@zname,
                      x = from@x,
                      y = from@y,
                      z = from@z,
                      pkg = from@pkg,
                      func = from@transform))
  if (length(from@x) && ! from@usetransform)
    return(data.frame(xname = from@xname,
                      yname = from@yname,
                      zname = from@zname,
                      x = from@x,
                      y = from@y,
                      z = from@z,
                      pkg = from@pkg,
                      refcode = from@refcode))
  data.frame(xname = character(),
             yname = character(),
             zname = character(),
             x = numeric(),
             y = numeric(),
             z = numeric(),
             pkg = character(),
             refcode = character())
})

#' @export
as.data.frame.xyz <-
  function(x, row.names = NULL, optional = FALSE, ...) as(x, "data.frame")

